demonstrations/tutorial_quantum_volume.py

```diff
@@ -0,0 +1,732 @@
+# uses *random* square circuits with a very particular form:
+#
+# .. figure:: ../demonstrations/quantum_volume/model_circuit_cross.png
+#     :align: center
+#     :width: 60%
+#
+#     ..
+#
+#     A schematic of the random circuit structure used in the quantum volume protocol.
+#     Image source: [#cross]_.
+#
+# Specifically, the circuits consist of :math:`d` sequential layers acting on
+# :math:`d` qubits. Each layer consists of two parts: a random permutation of
+# the qubits, followed by Haar-random SU(4) operations performed on neighbouring
+# pairs of qubits. (When the number of qubits is odd, the bottom-most qubit is
+# idle while the SU(4) operations run on the pairs. However, it will still be
+# incorporated by way of the permutations.) These circuits satisfy the criteria
+# in item 1 --- they have well-defined structure, and it is clear how they can be
+# scaled to different sizes.
+#
+# As for the compilation rules of item 2, to compute quantum volume we're
+# allowed to do essentially anything we'd like to the circuits in order to
+# improve them. This includes optimization, hardware-aware considerations such
+# as qubit placement and routing, and even resynthesis by finding unitaries that
+# are close to the target, but easier to implement on the hardware [#cross]_.
+#
+# Both the circuit structure and the compilation highlight how quantum volume is
+# about more than just the number of qubits. The error rates will affect the
+# achievable depth, and the qubit connectivity contributes through the layers of
+# permutations because a very well-connected processor will be able to implement
+# these in fewer steps than a less-connected one. Even the quality of the
+# software and the compiler plays a role here: higher-quality compilers will
+# produce circuits that fit better on the target devices, and will thus produce
+# higher quality results.
+#
+# The measures of success
+# ~~~~~~~~~~~~~~~~~~~~~~~
+#
+# Now that we have our circuits, we have to define the quantities that will
+# indicate how well we're able to run them. For that, we need a problem
+# to solve. The problem used for computing quantum volume is called the *heavy output
+# generation problem*. It has roots in the proposals for demonstrating quantum
+# advantage [#aaronson]_. Many such proposals make use of the properties of
+# various random quantum circuit families, as the distribution of the
+# measurement outcomes may not be easy to sample using classical
+# techniques.
+#
+# A distribution that is theorized to fulfill this property is the distribution
+# of *heavy* output bit strings. Heavy bit strings are those whose outcome
+# probabilities are above the median of the distribution. For example, suppose
+# we run a two-qubit circuit, and find that the measurement probabilities for
+# the output states are as follows:
+
+measurement_probs = {"00": 0.558, "01": 0.182, "10": 0.234, "11": 0.026}
```

Comment:
Might be nice to explicitly show the median calculation?
```suggestion
import numpy as np
measurement_probs = {"00": 0.558, "01": 0.182, "10": 0.234, "11": 0.026}
prob_array = np.fromiter(measurement_probs.values(), dtype=np.float)
np.median(prob_array)
```

Reply:
Good call!